ground/src/main/java/com/google/android/ground/repository/OfflineAreaRepository.kt

```diff
@@ -147,13 +152,31 @@ constructor(
 
   suspend fun hasHiResImagery(bounds: Bounds): Boolean {
     val client = getMogClient()
-    val maxZoom = client.collection.maxZoom
-    return client.buildTilesRequests(bounds.toGoogleMapsObject(), maxZoom..maxZoom).isNotEmpty()
+    val maxZoom = client.collection.sources.maxZoom()
+    return client.buildTilesRequests(bounds, maxZoom..maxZoom).isNotEmpty()
   }
 
   suspend fun estimateSizeOnDisk(bounds: Bounds): Int {
     val client = getMogClient()
-    val requests = client.buildTilesRequests(bounds.toGoogleMapsObject())
+    val requests = client.buildTilesRequests(bounds)
     return requests.sumOf { it.totalBytes }
   }
+
+  suspend fun sizeOnDevice(offlineArea: OfflineArea): Int =
+    offlineArea.tiles.sumOf { File(getLocalTileSourcePath(), it.getTilePath()).length().toInt() }
+
+  suspend fun removeFromDevice(offlineArea: OfflineArea) {
+    val tilesInSelectedArea = offlineArea.tiles
+    localOfflineAreaStore.deleteOfflineArea(offlineArea.id)
+    val remainingAreas = localOfflineAreaStore.offlineAreasOnceAndStream().awaitFirst()
+    val remainingTiles = remainingAreas.flatMap { it.tiles }.toSet()
+    val tilesToRemove = tilesInSelectedArea - remainingTiles
+    val tileSourcePath = getLocalTileSourcePath()
+    tilesToRemove.forEach {
+      val tilePath = File(tileSourcePath, it.getTilePath())
+      tilePath.delete()
+      tilePath.parentFile.deleteIfEmpty()
+      tilePath.parentFile?.parentFile.deleteIfEmpty()
+    }
+  }
```

Comment:
Good catch, thanks! Adding ktdoc now.

Afaict this logic behaves correctly with empty lists.. Did I miss something?

Reply:
Is it possible `tilesInSelectedArea` returns an empty list but `remainingTiles` is non-empty?

Reply:
We should never have an area with no times, but if it did what do you think might break?

Reply:
I'm not sure if anything would break, but this seems to be an invalid state which we should handle by throwing an exception. I don't feel too strongly about it but just wanted to understand if we have thought about it or not.

Reply:
The method guarantees there are no unused tiles left from the deleted area; if the area has no tiles then technically that result is achieved, so don't think we need to report an error here. Maybe we can log the odd state? Adding something.